proposals/infra/INF-0002-root-signature-in-clang.md

```diff
@@ -0,0 +1,717 @@
+
+The AST node serialization code might be looked like this:
+```
+void ASTDeclWriter::VisitHLSLRootSignatureDecl(HLSLRootSignatureDecl *D) {
+  VisitNamedDecl(D);
+  Record.AddSourceLocation(D->getKeywordLoc());
+  hlsl::ParsedRootSignature &RS = D->getRootSignature();
+  Record.push_back(RS.RSDesc.Version);
+  Record.push_back(RS.RSDesc.Flags);
+  
+  Record.push_back(ParsedRS.RSParameters.size());
+  for (const auto &P : ParsedRS.RSParameters) {
+    // push_back all the fields of root parameters.
+  }
+  Record.push_back(ParsedRS.StaticSamplers.size());
+  for (const auto &SS : ParsedRS.StaticSamplers) {
+    // push_back all the fields of static samplers.
+  }
+  
+  Code = serialization::DECL_HLSL_ROOT_SIGNATURE;
+}
+
+```
+
+
+A HLSLRootSignatureAttr will be created when meet RootSignature attribute in
+HLSL.
+
+```
+
+[RootSignature(MyRS1)]
+float4 main(float4 coord : COORD) : SV_Target
+{
+…
+}
+
+```
+
+Because the RootSignature attribute in hlsl only have the string, it is easier
+to add a StringArgument to save the string first.
+A HLSLRootSignatureDecl will be created and added to the HLSLRootSignatureAttr 
+for diagnostic and saved to the HLSLEntryRootSignatureAttr for clang 
+code generation.
+The HLSLRootSignatureDecl will save StringLiteral instead StringRef for 
+diagnostic.
+
+In clang code generation, the HLSLRootSignatureAttr in AST will be translated
+into metadata to express the layout and things like static sampler, root 
+flags, space and NumDescriptors in LLVM IR.
+
+CGHLSLRuntime will generate metadata to link the metadata as root
+signature for given entry function. 
+
+For case compile to a standalone root signature blob with option like:
+```
+
+dxc.exe /T rootsig_1_1 MyRS1.hlsl /E RS /Fo MyRS1.fxo
+```
+ the HLSLRootSignatureAttr will be bind to a fake empty entry.
+
+
+When In the case of local or global root signatures like:
+```
+
+GlobalRootSignature MyGlobalRootSignature =
+{
+    "DescriptorTable(UAV(u0)),"                     // Output texture
+    "SRV(t0),"                                      // Acceleration structure
+    "CBV(b0),"                                      // Scene constants
+    "DescriptorTable(SRV(t1, numDescriptors = 2))"  // Static index and vertex buffers.
+};
+
+LocalRootSignature MyLocalRootSignature = 
+{
+    "RootConstants(num32BitConstants = 4, b1)"  // Cube constants 
+};
+
+```
+ only a HLSLRootSignatureDecl will be created.
+
+Additionally, a new AST node, SubobjectToExportsAssociationDecl, should be 
+added to handle the SubobjectToExportsAssociation statement in HLSL.
+```
+SubobjectToExportsAssociation MyLocalRootSignatureAssociation =
+{
+    "MyLocalRootSignature",    // Subobject name
+    "MyHitGroup;MyMissShader"  // Exports association 
+};
+```
+The SubobjectToExportsAssociationDecl will contain a HLSLRootSignatureDecl and 
+a list of FunctionDecls.
+
+
+The default root signature version number is 2 which maps to rootsig_1_1.
```

Comment:
Seems like this should be higher up in the document.  Talking about the different versions of root signatures.  Are you calling this info out here because it is going to be written in the metadata at some point?

Also would the version also be stored in the deserialized version of the root signature?

Reply:
Cannot find a good place to add it. Do you have suggestion about it?

The version stored in deserialized/serialized version of root signature.